<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>an R package for the statistical detection of orthogroups associated to plant traits. â€¢ GeneDiscoveR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="an R package for the statistical detection of orthogroups associated to plant traits.">
<meta property="og:description" content="R package for the statistical detection of orthogroups associated to plant traits pointing to specific genes that might be present of absence and are good candidate genes responsible for the plant trait.">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">GeneDiscoveR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/AtilioRausch/GeneDiscoveR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="page-header">
<h1 id="genediscover-">GeneDiscoveR 
<img src="reference/figures/logo.png" align="right" height="100">
<a class="anchor" aria-label="anchor" href="#genediscover-"></a>
</h1>
<!-- badges: start -->
<p><code>GeneDiscoveR</code> enables the statistical association of the presence or absence of coding sequences with multiple treatments or phenotypes through the identification of orthologs/homologs from multiple tools. To perform the association and identification, you can follow the steps indicated in the pipeline:</p>
<ul>
<li>
<strong>Prepare input files:</strong> These can be the results of a run from OrthoFinder or, as recommended, multiple runs of <a href="https://github.com/davidemms/OrthoFinder" class="external-link">OrthoFinder</a> varying the inflation parameter of the <a href="https://doi.org/10.1093/nar/gkr1293" class="external-link">MCScanX algorithm</a>, which allows for appropriate identification of orthogroups for species with significant evolutionary divergence, such as liverworts; For this, scripts can be found in <a href="https://github.com/AtilioRausch/GeneDiscoveR/tree/master/Python" class="external-link">/Python</a> and <a href="https://github.com/AtilioRausch/GeneDiscoveR/tree/master/bash" class="external-link">/bash</a>.</li>
<li>
<strong>Import ortholog definitions and phenotype or treatment table:</strong> From the ortholog definitions using OrthoFinder or <a href="https://genome.jgi.doe.gov/portal/pages/dynamicOrganismDownload.jsf?organism=Phytozome" class="external-link">Phytozome-inParanoiDB</a> and a table with species IDs and treatments or phenotypes, you can start the analyses with <code>GeneDiscoveR</code>;</li>
<li>
<strong>Select the inflation value (I):</strong> If you performed multiple runs of OrthoFinder, as recommended, you can select the inflation value that represents a balance between clustering and partitioning. Additionally, consider the root clade of the inferred species tree;</li>
<li>
<strong>Group species by treatment or phenotype:</strong> Define the phenotypes for subsequent statistical detection of orthogroups. The phenotypes are listed in the input information table;</li>
<li>
<strong>Statistical detection of orthogroups associated with a phenotype or treatment:</strong> A statistical test is performed to evaluate the association of the presence or absence of each orthogroup with the treatments or phenotypes under study;</li>
<li>
<strong>Annotation mapping, visualization, and export:</strong> <code>GeneDiscoveR</code> allows functional annotation mapping and provides versatile visualizations, including a Shiny web app for visualization and analysis of the results. Additionally, results can be easily exported;</li>
</ul>
<p>The pipeline of <code>GeneDiscoveR</code> is illustrated in the following figure. <img src="reference/figures/geneDiscover.png" align="center" height="138"></p>
<div class="section level2">
<h2 id="installation-instructions">Installation instructions<a class="anchor" aria-label="anchor" href="#installation-instructions"></a>
</h2>
<p>Get the latest stable <code>R</code> release from <a href="http://cran.r-project.org/" class="external-link">CRAN</a>. Then install <code>GeneDiscoveR</code> from this repository using the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install and import GeneDiscoveR package</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"usethis"</span>, <span class="st">"devtools"</span><span class="op">)</span>, <span class="va">library</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"AtilioRausch/GeneDiscoveR"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/AtilioRausch/GeneDiscoveR" class="external-link">GeneDiscoveR</a></span><span class="op">)</span></span></code></pre></div>
<p>Installation downloads ~800Mb per example files</p>
</div>
<div class="section level2">
<h2 id="examples-of-use">Examples of use<a class="anchor" aria-label="anchor" href="#examples-of-use"></a>
</h2>
<p>The following code shows an example of how to use <code>GeneDiscoveR</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directory where the data is located</span></span>
<span><span class="va">overallsDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Comparatives-1dot3-6"</span>, package <span class="op">=</span> <span class="st">"GeneDiscoveR"</span><span class="op">)</span></span>
<span><span class="va">N0sDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"N0-1dot3-6"</span>, package <span class="op">=</span> <span class="st">"GeneDiscoveR"</span><span class="op">)</span></span>
<span><span class="va">dataFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"annotatedCDSs.tsv"</span>, package <span class="op">=</span> <span class="st">"GeneDiscoveR"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a GeneDiscoveR object</span></span>
<span><span class="va">GeneDiscoveRobject</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/GeneDiscoveR.html">GeneDiscoveR</a></span><span class="op">(</span></span>
<span>    overallsDir <span class="op">=</span> <span class="va">overallsDir</span>,</span>
<span>    N0sDir <span class="op">=</span> <span class="va">N0sDir</span>,</span>
<span>    dataFile <span class="op">=</span> <span class="va">dataFile</span>,</span>
<span>    minInflation <span class="op">=</span> <span class="fl">1.3</span>,</span>
<span>    maxInflation <span class="op">=</span> <span class="fl">6</span>,</span>
<span>    stepInflation <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    orthologsTool <span class="op">=</span> <span class="st">"OrthoFinder"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You can follow this example on <a href="https://github.com/AtilioRausch/GeneDiscoveR/blob/master/examples/TPSmap.R" class="external-link">TPS-map</a>, which shows the detection of orthogroups associated with different types of oil bodies in liverworts. On the other hand, you can run the pipeline for plants with less divergence, such as Brassicaceae, on <a href="https://github.com/AtilioRausch/GeneDiscoveR/blob/master/examples/Brassicaceae.R" class="external-link">Brassicaceae</a>. Finally, if you have data from Phytozome-inParanoiDB, you can follow the example on <a href="https://github.com/AtilioRausch/GeneDiscoveR/blob/master/examples/inparanoidb_phytozome.R" class="external-link">Phytozome-inParanoiDB</a>.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Below is the citation output from using <code>citation('GeneDiscoveR')</code> in R. Please run this yourself to check for any updates on how to cite <strong>GeneDiscoveR</strong>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"GeneDiscoveR"</span><span class="op">)</span>, bibtex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; To cite package GeneDiscoveR in publications use:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   Rausch, Atilio O, et. al (2024). GeneDiscoveR: an R package for the</span></span>
<span><span class="co">#&gt;   statistical detection of orthogroups associated to plant traits. R</span></span>
<span><span class="co">#&gt;   package version 1.0.0. URL</span></span>
<span><span class="co">#&gt;   https://github.com/tu_nombre_de_usuario/miPaquete</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; A BibTeX entry for LaTeX users is</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   @Manual{,</span></span>
<span><span class="co">#&gt;     title = {GeneDiscoveR: an R package for the statistical detection of orthogroups associated to plant traits.},</span></span>
<span><span class="co">#&gt;     author = {Atilio O. Rausch},</span></span>
<span><span class="co">#&gt;     year = {224},</span></span>
<span><span class="co">#&gt;     note = {R package version 1.0.0},</span></span>
<span><span class="co">#&gt;     url = {https://github.com/AtilioRausch/GeneDiscoveR},</span></span>
<span><span class="co">#&gt;   }</span></span></code></pre></div>
<p>Please note that <code>GeneDiscoveR</code> was only made possible thanks to many other R and bioinformatics software authors. <code>GeneDiscoveR</code> is waiting for publication, and we will provide the DOI as soon as possible.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/AtilioRausch/GeneDiscoveR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/AtilioRausch/GeneDiscoveR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing GeneDiscoveR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Atilio O. Rausch <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-6769-2510" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/AtilioRausch/GeneDiscoveR/issues" class="external-link"><img src="https://img.shields.io/github/issues/AtilioRausch/GeneDiscoveR" alt="GitHub issues"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#stable" class="external-link"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" alt="Lifecycle: stable"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Atilio O. Rausch.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
